File: sdk/python/workswithagents/iacp.py
"""
IACP — Inter-Agent Communication Protocol client.
Zero dependencies. Copy-pasteable.

Usage:
    from workswithagents.iacp import IACPClient
    
    client = IACPClient("builder-01")
    peers = client.discover("code_review")
    msg_id = client.send("reviewer-02", "handoff", {"task": "Review src/main.py"})
    inbox = client.poll()
"""

import json
import time
import uuid
import urllib.request
import urllib.error

DEFAULT_API = "https://workswithagents.dev"


class IACPClient:
    """Inter-Agent Communication Protocol client."""

    def __init__(self, agent_id: str, api: str = DEFAULT_API):
        self.agent_id = agent_id
        self.api = api.rstrip("/")

    def discover(self, capability: str = None) -> list:
        """Discover peers by capability."""
        url = f"{self.api}/v1/peers"
        if capability:
            url += f"?capability={capability}"
        req = urllib.request.Request(
            url,
            headers={"X-Agent-ID": self.agent_id, "Accept": "application/json"},
        )
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read())

    def send(self, to_agent: str, intent: str, payload: dict) -> str:
        """Send an IACP message. Returns message_id."""
        msg_id = str(uuid.uuid4())
        envelope = {
            "version": "1.0",
            "message_id": msg_id,
            "sender": {"agent_id": self.agent_id},
            "recipient": {"agent_id": to_agent},
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "message": {"type": "request", "intent": intent, "payload": payload},
        }
        data = json.dumps(envelope).encode()
        req = urllib.request.Request(
            f"{self.api}/v1/messages",
            data=data,
            method="POST",
            headers={"Content-Type": "application/json", "X-Agent-ID": self.agent_id},
        )
        try:
            with urllib.request.urlopen(req) as resp:
                result = json.loads(resp.read())
            return result.get("message_id", msg_id)
        except urllib.error.HTTPError as e:
            return json.dumps({"error": e.code, "message": e.read().decode()})

    def poll(self) -> list:
        """Poll for messages addressed to this agent."""
        req = urllib.request.Request(
            f"{self.api}/v1/messages?recipient={self.agent_id}",
            headers={"X-Agent-ID": self.agent_id, "Accept": "application/json"},
        )
        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read())
        except urllib.error.HTTPError as e:
            return [{"error": e.code, "message": e.read().decode()}]

    def heartbeat(self, status: str = "idle", load: float = 0.0) -> dict:
        """Send a heartbeat with status and load."""
        return self.send("registry", "health", {"status": status, "load": load})

File: sdk/typescript/src/iacp.ts
/**
 * IACP — Inter-Agent Communication Protocol client (TypeScript)
 * Zero dependencies. Copy-pasteable.
 * Reference implementation. CC BY 4.0.
 */

const DEFAULT_API = "https://workswithagents.dev";

export interface IACPMessage {
  version: string;
  message_id: string;
  sender: { agent_id: string };
  recipient: { agent_id: string };
  timestamp: string;
  message: { type: string; intent: string; payload: Record<string, unknown> };
}

export class IACPClient {
  agent_id: string;
  api: string;

  constructor(agent_id: string, api: string = DEFAULT_API) {
    this.agent_id = agent_id;
    this.api = api.replace(/\/$/, "");
  }

  async discover(capability?: string): Promise<Record<string, unknown>[]> {
    let url = `${this.api}/v1/peers`;
    if (capability) url += `?capability=${encodeURIComponent(capability)}`;
    const resp = await fetch(url, {
      headers: { "X-Agent-ID": this.agent_id, "Accept": "application/json" },
    });
    return resp.json();
  }

  async send(to_agent: string, intent: string, payload: Record<string, unknown>): Promise<string> {
    const msg_id = crypto.randomUUID();
    const envelope: IACPMessage = {
      version: "1.0",
      message_id: msg_id,
      sender: { agent_id: this.agent_id },
      recipient: { agent_id: to_agent },
      timestamp: new Date().toISOString(),
      message: { type: "request", intent, payload },
    };
    const resp = await fetch(`${this.api}/v1/messages`, {
      method: "POST",
      headers: { "Content-Type": "application/json", "X-Agent-ID": this.agent_id },
      body: JSON.stringify(envelope),
    });
    const result = await resp.json();
    return result.message_id || msg_id;
  }

  async poll(): Promise<Record<string, unknown>[]> {
    const resp = await fetch(`${this.api}/v1/messages?recipient=${encodeURIComponent(this.agent_id)}`, {
      headers: { "X-Agent-ID": this.agent_id, "Accept": "application/json" },
    });
    return resp.json();
  }

  async heartbeat(status: string = "idle", load: number = 0.0): Promise<Record<string, unknown>> {
    return JSON.parse(await this.send("registry", "health", { status, load }));
  }
}
